[Data] Add `pin_memory` to `iter_torch_batches`

srinathk10 opened the pull request:
Add `pin_memory` to `iter_torch_batches`
- Whether to pin the memory of the tensors. Defaults to False.

srinathk10 committed:
Add `pin_memory` to `iter_torch_batches`

Signed-off-by: Srinath Krishnamachari <srinath.krishnamachari@anyscale.com>

srinathk10 committed:
Fixes

Signed-off-by: Srinath Krishnamachari <srinath.krishnamachari@anyscale.com>

srinathk10 committed:
lint

Signed-off-by: Srinath Krishnamachari <srinath.krishnamachari@anyscale.com>

srinathk10 committed:
Revert "lint"

This reverts commit 467aa180085c4526ab5d6ec0f1360812cebe744e.

Signed-off-by: Srinath Krishnamachari <srinath.krishnamachari@anyscale.com>

srinathk10 committed:
Fixes

Signed-off-by: Srinath Krishnamachari <srinath.krishnamachari@anyscale.com>

srinathk10 committed:
Merge branch 'master' into srinathk10/iter-torch-batches-pinning

srinathk10 committed:
Fix Tests

Signed-off-by: Srinath Krishnamachari <srinath.krishnamachari@anyscale.com>

srinathk10 commented on python/ray/air/_internal/torch_utils.py:
```diff
@@ -461,6 +483,7 @@ def move_tensors_to_device(
     batch: TensorBatchType,
     device: Optional[Union[str, "torch.device"]] = None,
     non_blocking: bool = False,
```
Should we remove `non_blocking` flag here, since `pin_memory` flag can be used to decide whether to do `non_blocking` transfer?

raulchen replied:
yes.
it's not configurable by users, right?

justinvyu replied:
I think we can always default to `non_blocking=True` regardless of whether the tensor is pinned or not.

> Generally, tensor.to(device, non_blocking=True) is an effective choice for enhancing transfer speed.

Refer to https://docs.pytorch.org/tutorials/intermediate/pinmem_nonblock.html#non-blocking-true

raulchen commented on python/ray/air/_internal/torch_utils.py:
```diff
@@ -420,12 +434,20 @@ def concat_tensors_to_device(
+        (total_rows, *shape_tail),
+        dtype=dtype,
+        device=device,
+        pin_memory=pin_memory and device.type == "cpu",
```
`device.type` -> device can also be None or a string.

raulchen commented on python/ray/air/_internal/torch_utils.py:
```diff
@@ -374,11 +374,24 @@ def arrow_batch_to_tensors(
 
 
+def tensor_pin_memory_if_needed(tensor: torch.Tensor, pin_memory: bool) -> torch.Tensor:
+    if pin_memory and tensor.device.type == "cpu" and not tensor.is_pinned():
```
I remember checking `tensor.is_pinned()` for each tensor is expensive. 
not entirely sure, we can do a microbenchmark to confirm
but if so, we should remove it.

raulchen commented on python/ray/data/iterator.py:
```diff
@@ -379,6 +383,13 @@ def iter_torch_batches(
+                "pin_memory is set to True, but CUDA is not available. "
+                "pin_memory will be ignored."
+            )
+            pin_memory = False
```
actually, I think we can remove this.
it's useful when users are using other accelerators as well.

justinvyu commented on python/ray/air/_internal/torch_utils.py:
```diff
@@ -481,26 +504,37 @@ def move_tensors_to_device(
     if device is None:
         return batch
 
+    def to_device(tensor: torch.Tensor) -> torch.Tensor:
+        tensor = tensor_pin_memory_if_needed(tensor, pin_memory)
+
+        return tensor.to(
+            device, non_blocking=is_tensor_non_blocking_transfer(tensor, non_blocking)
+        )
```
I think pinning memory should be separated into a different stage of [the iterator pipeline](https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/block_batching/iter_batches.py#L123) -- after collate and before finalize (to).

Otherwise:

> Using `tensor.pin_memory().to(device, non_blocking=True)` can be up to twice as slow as a straightforward tensor.to(device).

Refer to https://docs.pytorch.org/tutorials/intermediate/pinmem_nonblock.html#pin-memory and https://docs.pytorch.org/tutorials/intermediate/pinmem_nonblock.html#additional-considerations

`tensor.pin_memory().to(device, non_blocking=True)` is what `move_tensors_to_device(..., pin_memory=True)` does during the finalize stage.

justinvyu commented on python/ray/air/_internal/torch_utils.py:
```diff
@@ -481,26 +504,37 @@ def move_tensors_to_device(
         return type(batch)(
-            [concat_tensors_to_device(t, device, non_blocking) for t in batch]
+            [
+                concat_tensors_to_device(t, device, non_blocking, pin_memory)
+                for t in batch
+            ]
         )
     elif _is_tensor_mapping(batch):
-        return {k: t.to(device, non_blocking=non_blocking) for k, t in batch.items()}
+        return {k: to_device(t) for k, t in batch.items()}
```
Refer to tensordict implementation, which performs the pin_memory for multiple with a threadpool so that multiple keys / elements can be pinned concurrently.

justinvyu reviewed (commented):
Some comments based on this torch tutorial: https://docs.pytorch.org/tutorials/intermediate/pinmem_nonblock.html

srinathk10 committed:
Misc fixes

Signed-off-by: Srinath Krishnamachari <srinath.krishnamachari@anyscale.com>

srinathk10 committed:
Merge branch 'master' into srinathk10/iter-torch-batches-pinning

Signed-off-by: srinathk10 <68668616+srinathk10@users.noreply.github.com>

srinathk10 committed:
Misc Fixes

Signed-off-by: Srinath Krishnamachari <srinath.krishnamachari@anyscale.com>

srinathk10 committed:
Merge branch 'master' into srinathk10/iter-torch-batches-pinning

srinathk10 committed:
Misc Fixes

Signed-off-by: Srinath Krishnamachari <srinath.krishnamachari@anyscale.com>

srinathk10 committed:
Misc Fixes

Signed-off-by: Srinath Krishnamachari <srinath.krishnamachari@anyscale.com>

srinathk10 committed:
Merge branch 'master' into srinathk10/iter-torch-batches-pinning

srinathk10 committed:
Lint

Signed-off-by: Srinath Krishnamachari <srinath.krishnamachari@anyscale.com>

srinathk10 committed:
Lint

Signed-off-by: Srinath Krishnamachari <srinath.krishnamachari@anyscale.com>

srinathk10 committed:
Release test fixes

Signed-off-by: Srinath Krishnamachari <srinath.krishnamachari@anyscale.com>

srinathk10 committed:
Merge branch 'master' into srinathk10/iter-torch-batches-pinning

raulchen commented on python/ray/data/collate_fn.py:
```diff
@@ -244,6 +255,10 @@ def __init__(
+        self.pin_memory = pin_memory
+
+        # Create thread pool executor for parallel tensor conversion
+        self._executor = ThreadPoolExecutor(max_workers=DEFAULT_COLLATE_FN_NUM_WORKERS)
```
does this help? `collate_fn` is already running in a multi-threading manner. 
a bit weird to have nested thread-pools. 
I'm wondering if we should just use more threads in the top-level thread pool - maybe decouple num threads with num prefetch batches.

srinathk10 replied:
Great point. 

- `collate_fn` is called as the last step after ray.get(), batching and formatting. Both batching and formatting are sequential and single threaded today.
- Given the abstraction of `collate_fn`, unit of concurrency can only be at batch granularity and so it's best to multi-thread in the collate_fn itself.

Let me write a micro-benchmark to see if threading helps. Release tests currently don't have enough columns to benefit from threading.

srinathk10 committed:
Addressed review comments

Signed-off-by: Srinath Krishnamachari <srinath.krishnamachari@anyscale.com>

srinathk10 committed:
Merge branch 'master' into srinathk10/iter-torch-batches-pinning

srinathk10 committed:
Merge branch 'master' into srinathk10/iter-torch-batches-pinning

srinathk10 commented:
Adding more threads helped for some batch sizes like 1024, 2048 but not uniformly. 


```
import os
import sys
import time
import ray
import numpy as np
import subprocess
from ray.data.context import DataContext

try:
    from tabulate import tabulate
    _HAS_TABULATE = True
except ImportError:
    _HAS_TABULATE = False

# Settings
DATASET_SIZE_BYTES = 1 * 1024 ** 3  # 1GB
ROW_SIZE_BYTES = 4 * 100  # 100 float32 columns per row
NUM_ROWS = DATASET_SIZE_BYTES // ROW_SIZE_BYTES
NUM_COLS = 100
BATCH_SIZES = [32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384]
NUM_WORKERS_LIST = [1, 2, 4, 8, 16, 32]


def run_benchmark(batch_size):
    ray.init(ignore_reinit_error=True)
    print(f"Generating dataset with {NUM_ROWS} rows, {NUM_COLS} float32 columns (~1GB)...")
    arrays = [np.random.rand(NUM_ROWS // 100, NUM_COLS).astype(np.float32) for _ in range(100)]
    ds = ray.data.from_numpy_refs([ray.put(arr) for arr in arrays])
    ds = ds.repartition(32)
    print(ds)
    iterator = ds.iterator()
    start = time.perf_counter()
    num_rows = 0
    for batch in iterator.iter_torch_batches(batch_size=batch_size):
        if isinstance(batch, dict):
            batch_len = next(iter(batch.values())).shape[0]
        else:
            batch_len = batch.shape[0]
        num_rows += batch_len
    elapsed = time.perf_counter() - start
    throughput = num_rows / elapsed
    print(f"RESULT: {batch_size},{os.environ.get('RAY_DATA_DEFAULT_COLLATE_FN_NUM_WORKERS','1')},{elapsed},{throughput}")


def main():
    if len(sys.argv) > 1 and sys.argv[1] == "_run_bench":
        batch_size = int(sys.argv[2])
        run_benchmark(batch_size)
        return

    results = []
    script_path = os.path.abspath(__file__)
    for batch_size in BATCH_SIZES:
        for num_workers in NUM_WORKERS_LIST:
            print(f"Benchmarking: batch_size={batch_size}, num_workers={num_workers}...")
            env = os.environ.copy()
            env["RAY_DATA_DEFAULT_COLLATE_FN_NUM_WORKERS"] = str(num_workers)
            proc = subprocess.run([
                sys.executable, script_path, "_run_bench", str(batch_size)
            ], capture_output=True, text=True, env=env)
            # Parse result from stdout
            for line in proc.stdout.splitlines():
                if line.startswith("RESULT:"):
                    _, values = line.split(":", 1)
                    bsz, nw, elapsed, throughput = values.split(",")
                    results.append((int(bsz), int(nw), float(elapsed), float(throughput)))
            print(proc.stdout)
            if proc.stderr:
                print(proc.stderr, file=sys.stderr)

    # Print results table
    headers = ["Batch Size", "Num Workers", "Time (s)", "Rows/sec"]
    table = [
        [batch_size, num_workers, f"{elapsed:.2f}", f"{throughput:.2f}"]
        for batch_size, num_workers, elapsed, throughput in results
    ]
    print("\nResults:")
    if _HAS_TABULATE:
        print(tabulate(table, headers=headers, tablefmt="github"))
    else:
        print(f"{'Batch Size':>10} | {'Num Workers':>12} | {'Time (s)':>10} | {'Rows/sec':>12}")
        print("-" * 52)
        for row in table:
            print(f"{row[0]:>10} | {row[1]:>12} | {row[2]:>10} | {row[3]:>12}")
        print("\nInstall tabulate for prettier tables: pip install tabulate")

if __name__ == "__main__":
    main() 

```

```
+------------+-------------+----------+--------------+------------------------------+
| Batch Size | Num Workers | Time (s) | Rows/sec     | % Improvement over 1 Worker |
+------------+-------------+----------+--------------+------------------------------+
| 32         | 1           | 63.8     | 42071.7      | 0.00%                        |
| 32         | 2           | 63.93    | 41986.3      | -0.20%                       |
| 32         | 4           | 63.8     | 42071.1      | -0.00%                       |
| 32         | 8           | 63.95    | 41976.1      | -0.23%                       |
| 32         | 16          | 63.25    | 42436.5      | +0.87%                       |
| 32         | 32          | 63.38    | 42355.8      | +0.67%                       |
+------------+-------------+----------+--------------+------------------------------+
| 64         | 1           | 32.95    | 81474.3      | 0.00%                        |
| 64         | 2           | 33.19    | 80883.6      | -0.72%                       |
| 64         | 4           | 32.66    | 82198.3      | +0.89%                       |
| 64         | 8           | 32.79    | 81870.9      | +0.49%                       |
| 64         | 16          | 33.16    | 80943.6      | -0.65%                       |
| 64         | 32          | 32.65    | 82202.4      | +0.89%                       |
+------------+-------------+----------+--------------+------------------------------+
| 128        | 1           | 17.5     | 153349       | 0.00%                        |
| 128        | 2           | 17.22    | 155919       | +1.67%                       |
| 128        | 4           | 17.25    | 155634       | +1.49%                       |
| 128        | 8           | 17.05    | 157421       | +2.65%                       |
| 128        | 16          | 17.37    | 154538       | +0.78%                       |
| 128        | 32          | 17.18    | 156206       | +1.86%                       |
+------------+-------------+----------+--------------+------------------------------+
| 256        | 1           | 9.37     | 286334       | 0.00%                        |
| 256        | 2           | 9.5      | 282508       | -1.34%                       |
| 256        | 4           | 9.55     | 281185       | -1.80%                       |
| 256        | 8           | 9.46     | 283642       | -0.94%                       |
| 256        | 16          | 9.33     | 287726       | +0.49%                       |
| 256        | 32          | 9.57     | 280590       | -2.00%                       |
+------------+-------------+----------+--------------+------------------------------+
| 512        | 1           | 5.61     | 478275       | 0.00%                        |
| 512        | 2           | 5.6      | 479203       | +0.19%                       |
| 512        | 4           | 5.55     | 483426       | +1.08%                       |
| 512        | 8           | 5.52     | 486272       | +1.67%                       |
| 512        | 16          | 5.6      | 479584       | +0.27%                       |
| 512        | 32          | 5.61     | 478416       | +0.03%                       |
+------------+-------------+----------+--------------+------------------------------+
| 1024       | 1           | 3.7      | 725703       | 0.00%                        |
| 1024       | 2           | 3.68     | 728484       | +0.38%                       |
| 1024       | 4           | 3.63     | 738812       | +1.81%                       |
| 1024       | 8           | 3.64     | 738374       | +1.74%                       |
| 1024       | 16          | 3.63     | 739950       | +1.96%                       |
| 1024       | 32          | 3.58     | 749438       | +3.27%                       |
+------------+-------------+----------+--------------+------------------------------+
| 2048       | 1           | 2.71     | 992336       | 0.00%                        |
| 2048       | 2           | 2.71     | 992245       | -0.01%                       |
| 2048       | 4           | 2.71     | 989905       | -0.24%                       |
| 2048       | 8           | 2.61     | 1026520      | +3.45%                       |
| 2048       | 16          | 2.6      | 1034310      | +4.24%                       |
| 2048       | 32          | 2.65     | 1013810      | +2.17%                       |
+------------+-------------+----------+--------------+------------------------------+
| 4096       | 1           | 2.19     | 1224270      | 0.00%                        |
| 4096       | 2           | 2.22     | 1207820      | -1.34%                       |
| 4096       | 4           | 2.17     | 1236140      | +0.97%                       |
| 4096       | 8           | 2.19     | 1224840      | +0.05%                       |
| 4096       | 16          | 2.18     | 1230410      | +0.50%                       |
| 4096       | 32          | 2.16     | 1240010      | +1.29%                       |
+------------+-------------+----------+--------------+------------------------------+
| 8192       | 1           | 1.92     | 1399630      | 0.00%                        |
| 8192       | 2           | 2.02     | 1326850      | -5.19%                       |
| 8192       | 4           | 1.97     | 1365810      | -2.42%                       |
| 8192       | 8           | 1.95     | 1374570      | -1.79%                       |
| 8192       | 16          | 2.01     | 1336830      | -4.48%                       |
| 8192       | 32          | 1.94     | 1384000      | -1.11%                       |
+------------+-------------+----------+--------------+------------------------------+
| 16384      | 1           | 1.93     | 1393800      | 0.00%                        |
| 16384      | 2           | 1.93     | 1387270      | -0.47%                       |
| 16384      | 4           | 1.98     | 1357230      | -2.63%                       |
| 16384      | 8           | 1.92     | 1400560      | +0.49%                       |
| 16384      | 16          | 2.02     | 1329230      | -4.63%                       |
| 16384      | 32          | 1.91     | 1402560      | +0.63%                       |
+------------+-------------+----------+--------------+------------------------------+
```

srinathk10 commented:
PM (Pinned memory) helps only for batch size 32 and 1-worker.



```
| Batch | Workers |  nPM Rows/sec | nPM % |   PM Rows/sec | PM % |
|-------|---------|--------------:|------:|--------------:|-----:|
|   32  |    1    |     31412.6   | 89.2% |     32541.5   | 100% |
|   32  |    2    |     34725.3   | 98.5% |     32567.2   | 100% |
|   32  |    4    |     35062.1   | 99.5% |     32080.4   | 98.5%|
|   32  |    8    |     35049.3   | 99.5% |     32158.1   | 98.8%|
|   32  |   16    |     35249.4   |100.0% |     32300.1   | 99.2%|
|   32  |   32    |     35052.2   | 99.4% |     32295.3   | 99.2%|
|-------|---------|--------------:|------:|--------------:|-----:|
|   64  |    1    |     57961.1   |100.0% |     47588.7   | 98.7%|
|   64  |    2    |     57435.3   | 99.1% |     48146.8   |100.0%|
|   64  |    4    |     56947.6   | 98.2% |     48134.4   |100.0%|
|   64  |    8    |     56903.3   | 98.2% |     48250.4   |100.2%|
|   64  |   16    |     57378.9   | 99.0% |     48159.3   |100.0%|
|   64  |   32    |     57224.9   | 98.7% |     48073.1   | 99.8%|
|-------|---------|--------------:|------:|--------------:|-----:|
|  128  |    1    |     83979.3   | 98.2% |     70181.1   | 99.5%|
|  128  |    2    |     85064.0   | 99.5% |     69727.5   | 98.8%|
|  128  |    4    |     84949.3   | 99.4% |     70043.8   | 99.4%|
|  128  |    8    |     84160.8   | 98.5% |     70402.5   |100.0%|
|  128  |   16    |     85559.0   |100.0% |     70507.6   |100.1%|
|  128  |   32    |     84069.3   | 98.3% |     70042.8   | 99.4%|
|-------|---------|--------------:|------:|--------------:|-----:|
|  256  |    1    |    111611.0   | 99.6% |     90302.2   | 97.6%|
|  256  |    2    |    109557.0   | 97.8% |     90721.8   | 98.1%|
|  256  |    4    |    109822.0   | 98.0% |     90253.4   | 97.5%|
|  256  |    8    |    112152.0   |100.0% |     93131.9   |100.7%|
|  256  |   16    |    111337.0   | 99.3% |     91295.6   | 98.7%|
|  256  |   32    |    110086.0   | 98.2% |     91350.9   | 98.8%|
|-------|---------|--------------:|------:|--------------:|-----:|
|  512  |    1    |    130919.0   | 96.2% |    108348.0   | 98.4%|
|  512  |    2    |    130453.0   | 95.8% |    106796.0   | 97.0%|
|  512  |    4    |    131811.0   | 96.8% |    108625.0   | 98.6%|
|  512  |    8    |    136060.0   |100.0% |    110060.0   |100.0%|
|  512  |   16    |    131386.0   | 96.6% |    105619.0   | 96.0%|
|  512  |   32    |    128869.0   | 94.7% |    105998.0   | 96.3%|
|-------|---------|--------------:|------:|--------------:|-----:|
| 1024  |    1    |    139105.0   | 97.8% |    117632.0   |100.0%|
| 1024  |    2    |    141753.0   | 99.7% |    115784.0   | 98.4%|
| 1024  |    4    |    142246.0   |100.0% |    115910.0   | 98.5%|
| 1024  |    8    |    139860.0   | 98.3% |    111982.0   | 95.2%|
| 1024  |   16    |    139233.0   | 97.9% |    114884.0   | 97.7%|
| 1024  |   32    |    137309.0   | 96.5% |    116046.0   | 98.7%|
|-------|---------|--------------:|------:|--------------:|-----:|
| 2048  |    1    |    111963.0   | 98.0% |     97506.3   |100.0%|
| 2048  |    2    |    114805.0   |100.6% |     97299.3   | 99.8%|
| 2048  |    4    |    113229.0   | 99.5% |     97452.0   |100.0%|
| 2048  |    8    |    112422.0   | 98.8% |     98904.6   |101.4%|
| 2048  |   16    |    113609.0   | 99.7% |     97714.5   |100.2%|
| 2048  |   32    |    114222.0   |100.2% |     97030.7   | 99.5%|
|-------|---------|--------------:|------:|--------------:|-----:|
| 4096  |    1    |    110541.0   | 99.0% |     92419.3   | 99.0%|
| 4096  |    2    |    107532.0   | 96.4% |     92174.0   | 98.7%|
| 4096  |    4    |    109512.0   | 98.1% |     92235.2   | 98.8%|
| 4096  |    8    |    110144.0   | 98.7% |     92212.0   | 98.8%|
| 4096  |   16    |    109704.0   | 98.2% |     91793.1   | 98.2%|
| 4096  |   32    |    110647.0   |100.0% |     93335.8   |100.0%|
```

srinathk10 committed:
Addressed review comments

Signed-off-by: Srinath Krishnamachari <srinath.krishnamachari@anyscale.com>

srinathk10 committed:
Merge branch 'master' into srinathk10/iter-torch-batches-pinning

srinathk10 committed:
Misc Fixes

Signed-off-by: Srinath Krishnamachari <srinath.krishnamachari@anyscale.com>

srinathk10 committed:
Merge branch 'master' into srinathk10/iter-torch-batches-pinning

srinathk10 committed:
Lint

Signed-off-by: Srinath Krishnamachari <srinath.krishnamachari@anyscale.com>
